Jukebox

DEATHB4DEFEAT opened the pull request:
# Description

Adds a functional Jukebox to the stations! Capable of playing music, being paused, queueing up songs, and a whole load of other stuff the engine doesn't actually allow for yet ' v ,

---

# Tasks

- [x] Functional system
- [x] Fix the next song not playing automatically
- [x] Clean UI
- ~~[ ] Sync music over the network for new joins~~ Will probably require a pretty heavy rewrite, this is fine for now.
- [x] More songs
- [x] Construction
- [x] Research ~~Uses stock parts, that counts~~
- [x] Craftable boards
- [ ] Larger control icons
- [x] Animate sprite

---

<details><summary><h1>Media</h1></summary>
<p>

# [Youtube Demonstration](https://youtu.be/ogi6_eaFO8g)

https://github.com/Simple-Station/Parkstation/assets/81725545/4332b071-1c13-48f4-948d-5e3dcc854ab3

![Content Client_Er7gPqlj1S](https://github.com/Simple-Station/Parkstation/assets/77995199/d9a1b828-4c6f-49ae-b969-edb04ff20780)

![Content Client_1fDBkySTlq](https://github.com/Simple-Station/Parkstation/assets/77995199/f815a77a-9c12-482b-bc66-35eb482ba57b)

</p>
</details>

---

# Changelog

<!-- You can add an author after the `:cl:` to change the name that appears in the changelog, ex: `:cl: Death`, it will default to your GitHub display name -->

:cl: Pspritechologist & DEATHB4DEFEAT
- add: NanoTrasen have heard that employees function better when allowed to listen to music. They trapped some elves in a wooden box- And wouldn't you know it, it's *more* than just generated digital notes!*

Pspritechologist committed:
All of it

Pspritechologist committed:
Merge branch 'master' of https://github.com/Simple-Station/Parkstation into jukebox

DEATHB4DEFEAT committed:
it.. works?

Pspritechologist committed:
Removed borked scrolls, fixed trackbar

DEATHB4DEFEAT committed:
what

DEATHB4DEFEAT committed:
Merge branch 'master' of https://github.com/Park-station/Parkstation into jukebox

DEATHB4DEFEAT committed:
it's something?

DEATHB4DEFEAT committed:
locale

DEATHB4DEFEAT committed:
more queued songs

DEATHB4DEFEAT committed:
everything stylish

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,308 @@
+        {
+            Logger.Error($"Jukebox track prototype {song} not found!");
+
+            Skip(jukeBox, jukeboxComp);
```
This function won't *always* be called in order to play a queued song, so skipping isn't the right move here.

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,308 @@
+
+        jukeboxComp.FinishPlayingTime = _timing.CurTime + song.Duration - offset;
+
+        jukeboxComp.CurrentlyPlayingStream = _audio.Play(song.Path, Filter.Broadcast(), jukeBox, true, AudioParams.Default.WithPlayOffset((float)offset.TotalSeconds));
```
This variable should be managed on the client, and the Jukebox should simply sync the information that it *is* playing. The Client can then handle playing the song, calculating the offset by time remaining, time started, etc.. This way, new players will have their Jukeboxes synced with current players.
I also don't know how the current system will play with PVS.

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,308 @@
+    #endregion Public functions
+    #region Private functions
+
+    /// <summary>
+    ///     Sets the Jukebox to either a state where it can or cannot play songs.
+    /// </summary>
+    private void CheckCanPlay(EntityUid uid, JukeboxComponent component)
+    {
+        var canPlay = true;
+
+        if (EntityManager.TryGetComponent<ApcPowerReceiverComponent>(uid, out var powerComp) && !powerComp.Powered)
+            canPlay = false;
+
+        component.CanPlay = canPlay;
+
+        if (canPlay)
+            TryRestart(uid, component);
+        else
+            Stop(uid, component);
+    }
```
Should probably be public, honestly? I dunno, maybe if we added a way to state that Jukeboxes were disabled or something.

OH speaking of which EMPs need to do something. I'm thinking- would it be too much to stop the song, play it with the correct offset from where it was but add like random reverb to it or some shit? We can check the AudioParams, I think they got some funny stuff. After EMP duration, just stop it again and play like normal.

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,308 @@
+        if (jukeboxComp.Paused || !jukeboxComp.CanPlay)
+            return;
+
+        jukeboxComp.Playing = true;
+
+        if (jukeboxComp.CurrentlyPlayingTrack == null || jukeboxComp.FinishPlayingTime == null || jukeboxComp.StoppedTime == null)
+        {
+            UpdateState(jukeBox, jukeboxComp);
+            return;
+        }
```
Should playing be true even if it isn't playing?... Seems like a dumb question, but I actually think that may be how this works...

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
*All* the functions in here need to be renamed to be in line with convention, *and not just fucking stupuid*.
"End."
Yeah, thanks me.

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystems.Interactions.cs:
This file needs a better name, it's not all interactions. This file contains things seperate from the main logic of the Jukebox. System interactions, I suppose, but even that's iffy.
Also EMP should go here :)

Pspritechologist replied:
Eh

Pspritechologist commented on Content.Shared/SimpleStation14/Jukebox/JukeboxComponent.cs:
```diff
@@ -0,0 +1,131 @@
+    [DataField("maxQueued")] [ViewVariables(VVAccess.ReadWrite)]
+    public int MaxQueued { get; set; } = 7;
+
+    /// <summary>
+    ///     The song art to be used when no song is playing.
+    /// </summary>
+    [AutoNetworkedField]
+    [DataField("defaultSongArtPath")]
+    public string DefaultSongArtPath { get; set; } = "/Textures/SimpleStation14/JukeboxTracks/default.png";
```
Does this need to be networked?? How likely is it this is changing??
Confused.

Pspritechologist commented on Content.Shared/SimpleStation14/Jukebox/JukeboxComponent.cs:
```diff
@@ -0,0 +1,131 @@
+    ///     Whether or not the Jukebox should currently be playing.
+    /// </summary>
+    /// <remarks>
+    ///     This can still be true if no song is playing. It's essentially <see cref="Paused"/> but not controlled by the user.
```
Not only is this variable poorly named, *this remark isn't even useful.* It's extremely unclear how this variable functions, I believe it's simply "Is this Jukebox *allowed* to play," but I'd need to check over all this again.

Pspritechologist commented on Content.Shared/SimpleStation14/Prototypes/JukeboxTrackPrototype.cs:
```diff
@@ -0,0 +1,73 @@
+    [IdDataField]
+    public string ID { get; } = default!;
+
+    /// <summary>
+    ///     Input string for the name, pre-localization.
+    /// </summary>
+    [DataField("name")]
+    private string NameString { get; } = default!;
+
+    /// <summary>
+    ///     Actual name, localized.
+    /// </summary>
+    public string Name => Loc.GetString(NameString);
```
This is why everything is an Entity nowadays :P

Pspritechologist commented on Content.Shared/SimpleStation14/Prototypes/JukeboxTrackPrototype.cs:
```diff
@@ -0,0 +1,73 @@
+    [DataField("path")]
+    public SoundSpecifier Path { get; } = default!;
+
+    /// <summary>
+    ///     Serialized as a string because TimeSpan is not serializable, and converted to a TimeSpan with <see cref="ToTimeSpan"/>.
+    /// </summary>
+    [DataField("duration")]
+    private string DurationString { get; } = default!;
+
+    /// <summary>
+    ///     Actual duration as a TimeSpan.
+    /// </summary>
+    public TimeSpan Duration => ToTimeSpan(DurationString);
```
We should just learn how to make a serializer... I honestly don't think it'd be *that* hard.

Pspritechologist commented:
Also note; The system needs to react to the unpaused event so that the saved time is set correctly.

Pspritechologist committed:
Jukeboxes can now be emagged. Renamed a handful of things to be less dumb.

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,320 @@
+        jukeboxComp.CurrentlyPlayingStream = _audio.Play(song.Path, Filter.Broadcast(), jukeBox, true, AudioParams.Default.WithPlayOffset((float) offset.TotalSeconds));
+
+        Timer.Spawn((int) (song.Duration.TotalMilliseconds - offset.TotalMilliseconds), () => OnSongEnd(jukeBox, jukeboxComp), jukeboxComp.SongTimerCancel.Token);
+
```
Still don't know why this doesn't work.

Pspritechologist committed:
Handled map pausing, added some docs, and generally clarified things.

Felt like I did more :pensive:

Pspritechologist committed:
Added upgradable Jukebox parts (which means Jukeboxes are constructable now), removable decorative portions, lots of commenting, and some general clean up.

Also added a couple extra Jukebox examples in the YAML

Pspritechologist committed:
Replaced the timer with an update check. Autoplay is working baybee

Pspritechologist committed:
Jukebox now supports signalling, both for controlling it and reacting to its song state.

It also plays when you punch it.

Pspritechologist committed:
Several naming and doc clarifications- and Jukeboxes now have proper appearances!

Pspritechologist commented on Content.Shared/SimpleStation14/Jukebox/JukeboxComponent.cs:
```diff
@@ -0,0 +1,157 @@
+    [DataField("defaultSongArtPath")]
+    public string DefaultSongArtPath { get; set; } = "/Textures/SimpleStation14/JukeboxTracks/default.png";
+
+    /// <summary>
+    ///     A colour to be used in the Jukebox's UI.
+    ///     Should be based on the sprite.
+    /// </summary>
+    [DataField("jukeboxUiColorBG")]
+    public string JukeboxUiColorBG { get; set; } = "#602C00";
+
+    /// <inheritdoc cref="JukeboxUiColorBG"/>
+    [DataField("jukeboxUiColorPanel")]
+    public string JukeboxUiColorPanel { get; } = "#480F0F";
+
+    /// <inheritdoc cref="JukeboxUiColorBG"/>
+    [DataField("jukeboxUiColorAccent")]
+    public string JukeboxUiColorAccent { get; } = "#20181B";
```
These should be colors, as serializers for those exist.

Pspritechologist committed:
Tons of stuff.

Apart from adding a whack ton of songs, just loads of general clean up and additions before release.

Pspritechologist committed:
Component uses actual colors

Pspritechologist committed:
Uses ResPath now

Also slightly cleaner, a fix or two

Pspritechologist committed:
Danya Vodogonez

Danya Vodovoz is out, Dieter van der Westen is in. It was time for a change :)
Also adds a hand full new Jukebox tracks.

Pspritechologist commented:
@DEATHB4DEFEAT You gotta make nicer UI icons, and increase the size of the UI a bit. It needs a once over (is all of the text in it good to keep?), and then this is ready for prime time baybeeee

Pspritechologist commented on Content.Client/SimpleStation14/Jukebox/Ui/JukeboxSystem.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Content.Shared.SimpleStation14.Jukebox;
+
+public sealed class SharedJukeBoxSystem : EntitySystem
```
This is ClientSide, not Shared.

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,352 @@
+    public const string PortSkip = "Skip";
+    public const string PortPause = "Pause";
+    public const string PortUnPause = "Unpause";
+    public const string PortTogglePuase = "TogglePause";
```
puase

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,352 @@
+    ///     If the Jukebox is already playing a song, it will be stopped and replaced with the new song.
+    ///     To queue a song instead, see <see cref="TryQueueSong"/>.
+    /// </summary>
+    /// <param name="song">The JukeboxTrackPrototype representing the song to be played.</param>
```
Not a prototype

Pspritechologist commented on Content.Server/SimpleStation14/Jukebox/JukeboxSystem.cs:
```diff
@@ -0,0 +1,352 @@
+    [Dependency] private readonly DeviceLinkSystem _link = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+
+    public const string PortSongPlayed = "Start";
```
These really shouldn't be hardcoded I don't think, but making them variables makes checking them a pain...
